x/evm/keeper/integration_test.go

```diff
@@ -1,246 +1,252 @@
+	testkeyring "github.com/evmos/evmos/v14/testutil/integration/keyring"
+	"github.com/evmos/evmos/v14/testutil/integration/network"
+
+	abcitypes "github.com/cometbft/cometbft/abci/types"
 	ethtypes "github.com/ethereum/go-ethereum/core/types"
-	"github.com/evmos/evmos/v14/app"
-	"github.com/evmos/evmos/v14/crypto/ethsecp256k1"
-	"github.com/evmos/evmos/v14/encoding"
-	"github.com/evmos/evmos/v14/testutil"
-	utiltx "github.com/evmos/evmos/v14/testutil/tx"
-	"github.com/evmos/evmos/v14/utils"
-	"github.com/evmos/evmos/v14/x/feemarket/types"
-
-	dbm "github.com/cometbft/cometbft-db"
-	abci "github.com/cometbft/cometbft/abci/types"
-	"github.com/cometbft/cometbft/libs/log"
-	simutils "github.com/cosmos/cosmos-sdk/testutil/sims"
-	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
 	evmtypes "github.com/evmos/evmos/v14/x/evm/types"
 )
 
-var _ = Describe("Feemarket", func() {
-	var privKey *ethsecp256k1.PrivKey
+type IntegrationTestSuite struct {
+	network     network.Network
+	factory     factory.TxFactory
+	grpcHandler grpc.Handler
+	keyring     testkeyring.Keyring
+}
 
-	Describe("Performing EVM transactions", func() {
-		type txParams struct {
-			gasLimit  uint64
-			gasPrice  *big.Int
-			gasFeeCap *big.Int
-			gasTipCap *big.Int
-			accesses  *ethtypes.AccessList
+var _ = Describe("Handling a MsgEthereumTx message", Label("EVM"), Ordered, func() {
+	var s *IntegrationTestSuite
+
+	BeforeAll(func() {
+		keyring := testkeyring.New(3)
+		integrationNetwork := network.New(
+			network.WithPreFundedAccounts(keyring.GetAllAccAddrs()...),
+		)
+		grpcHandler := grpc.NewIntegrationHandler(integrationNetwork)
+		txFactory := factory.New(integrationNetwork, grpcHandler)
+		s = &IntegrationTestSuite{
+			network:     integrationNetwork,
+			factory:     txFactory,
+			grpcHandler: grpcHandler,
+			keyring:     keyring,
 		}
-		type getprices func() txParams
+	})
+
+	AfterEach(func() {
+		// Start each test with a fresh block
+		err := s.network.NextBlock()
+		Expect(err).To(BeNil())
+	})
+
+	When("the params have default values", Ordered, func() {
+		BeforeAll(func() {
+			// Set params to default values
+			defaultParams := evmtypes.DefaultParams()
+			err := s.network.UpdateEvmParams(defaultParams)
+			Expect(err).To(BeNil())
+
+			err = s.network.NextBlock()
+			Expect(err).To(BeNil())
+		})
 
-		Context("with MinGasPrices (feemarket param) < BaseFee (feemarket)", func() {
-			var (
-				baseFee      int64
-				minGasPrices int64
+		It("performs a transfer transaction", func() {
+			senderPriv := s.keyring.GetPrivKey(0)
+			receiver := s.keyring.GetKey(1)
+			txArgs := evmtypes.EvmTxArgs{
+				To:     &receiver.Addr,
+				Amount: big.NewInt(1000),
+			}
+
+			res, err := s.factory.ExecuteEthTx(senderPriv, txArgs)
+			Expect(err).To(BeNil())
+			Expect(res.IsOK()).To(Equal(true), "transaction should have succeeded", res.GetLog())
+		})
+
+		It("performs a contract deployment and contract call", func() {
```

Comment:
this test should be split in 2, one for contract deployment and one for contract call

Reply:
I do agree tests should be atomic, but the reason why this two calls are together is because they are dependent from the other. E.g if contract deployment fails contract call will fail and I am trying to avoid that on tests. Also the global variables we are use to having on tests which are very error prone. But happy to change it if you disagree